### ISSUE TITLE
Non-deterministic failures when running with pytest-xdist

### ISSUE THREAD

**Issue by spookylukey** [django-compressor/django-compressor#1103]
When running Django tests with pytest and pytest-xdist, I get occasional failures.

This extension runs tests in parallel, in multiple processes (which is very helpful due to large speedups) . Sometimes, but not always, it seems we get interference between the runs, and the following stacktrace:
```
/home/luke/.virtualenvs/rpr/lib/python3.9/site-packages/compressor/base.py:245: in hunks
    yield self.handle_output(kind, value, forced=True,
/home/luke/.virtualenvs/rpr/lib/python3.9/site-packages/compressor/base.py:338: in handle_output
    return output_func(mode, content, forced, basename)
/home/luke/.virtualenvs/rpr/lib/python3.9/site-packages/compressor/base.py:350: in output_file
    self.storage.save(new_filepath, ContentFile(content.encode(self.charset)))
/home/luke/.virtualenvs/rpr/lib/python3.9/site-packages/django/core/files/storage.py:54: in save
    return self._save(name, content)
/home/luke/.virtualenvs/rpr/lib/python3.9/site-packages/django/core/files/storage.py:298: in _save
    os.chmod(full_path, self.file_permissions_mode)
```

Resulting in:
```
FileNotFoundError: [Errno 2] No such file or directory: '/home/luke/.../static/CACHE/css/main.aaa6e4bfc1a0.css'
```
The line it is failing on:
```python
    
        if self.file_permissions_mode is not None:
            os.chmod(full_path, self.file_permissions_mode)
```

This looks very similar to  #1099 but I'm putting it in a different issue, because this seems more likely to be related to multi-process race conditions.

**Comment by diox** [comment]
I'm pretty sure this is the same underlying problem as https://github.com/django-compressor/django-compressor/issues/1099.

I believe that ultimately the problem is a race condition cause by the fact that compressor really wants to guarantee what the filename is going to be when writing files. Because django doesn't let it do that, it tries to delete the file first. Unfortunately multiple processes (or threads in #1099) are trying to write to the same file at the same time, and we end up with this race condition where sometimes the file is no longer present...

We're discussing a fix in https://github.com/django-compressor/django-compressor/pull/1100 but it's difficult without completely refactoring how compressor deals with files. We can avoid the problem for a single `compress` process using multiple threads, but it's going to be more difficult in the general case like this one which involves multiple processses.

**Comment by diox** [comment]
https://github.com/django-compressor/django-compressor/issues/1097 is probably the same thing as well.

**Comment by diox** [comment]
@spookylukey this should be fixed by https://github.com/django-compressor/django-compressor/commit/00910e0bd5295044befacdad7b88249ec7bbf142. Could you please test with the `develop` branch and report back ? Thanks

**Comment by diox** [comment]
This should be fixed in 4.0.

**Comment by spookylukey** [comment]
Thanks, I'm sorry that I wasn't able to test before.

I have a set of tests that I can use to pretty reliably reproduce the issue using pytest-xdist. Running 22 tests with concurrency of 3 processes, and repeating 10 times, I got the following run of failure counts due to this issue:

2 2 1 2 0 1 1 0 0 2

After upgrading to 4.0, I cannot get this failure at all, even cranking up concurrency to 8. So this is pretty good evidence it is now fixed for the multi-process case. Thanks so much!
